coffeecard/CoffeeCard.WebApi/Services/LoginLimiter.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Concurrent;
+using CoffeeCard.Common.Configuration;
+using CoffeeCard.WebApi.Models;
+using Serilog;
+
+namespace CoffeeCard.WebApi.Services
+{
+    public class LoginLimiter : ILoginLimiter
+    {
+        private readonly ConcurrentDictionary<string, (DateTime, int)> _loginAttempts;
+        private readonly IdentitySettings _identitySettings;
+
+        public LoginLimiter(IdentitySettings identitySettings)
+        {
+            _loginAttempts = new ConcurrentDictionary<string, (DateTime, int)>();
+            _identitySettings = identitySettings;
+        }
+
+        /// <summary>
+        /// Receives an email address to update the count for.
+        /// If no count currently exist for said email, creates a new one count with the current time.
+        /// If a count exist, increment the counter by 1 for said email, and preserve the time.
+        /// </summary>
+        /// <param name="email"></param>
+        /// <returns>
+        /// The value that the key was updated with
+        /// </returns>
+        private (DateTime, int) UpdateAndGetLoginAttemptCount(string email)
+        {
+            return _loginAttempts.AddOrUpdate(email, (DateTime.UtcNow, 0), (key, value) => (value.Item1, value.Item2 +1));
+        }
+
+        /// <summary>
+        /// Resets the first failed login time and in so doing, extends the lockout period
+        /// </summary>
+        /// <param name="email"></param>
+        private void ResetUsersTimeoutPeriod(string email)
+        {
+            if (!_loginAttempts.TryGetValue(email, out var oldEntry)) return;
```

Comment:
You could refactor this and use the pair just retrieved from the map on line 52

Reply:
Since it is a map the operation should only cost O(1) right? I think the code is a bit cleaner this way, but if you disagree I'll reformat